test/functional/multifactor_auths_controller_test.rb

```diff
@@ -82,19 +82,39 @@ class MultifactorAuthsControllerTest < ActionController::TestCase
 
       context 'on POST to create mfa' do
         setup do
-          seed = ROTP::Base32.random_base32
-          @controller.session[:mfa_seed] = seed
-          post :create, params: { otp: ROTP::TOTP.new(seed).now }
+          @seed = ROTP::Base32.random_base32
+          @controller.session[:mfa_seed] = @seed
         end
 
-        should respond_with :success
-        should 'show recovery codes' do
-          @user.reload.mfa_recovery_codes.each do |code|
-            assert page.has_content?(code)
+        context 'when qr-code is not expired' do
+          setup do
+            @controller.session[:mfa_seed_expire] = Gemcutter::MFA_KEY_EXPIRY.from_now.utc.to_i
+            post :create, params: { otp: ROTP::TOTP.new(@seed).now }
+          end
+
+          should respond_with :success
+          should 'show recovery codes' do
+            @user.reload.mfa_recovery_codes.each do |code|
+              assert page.has_content?(code)
+            end
+          end
+          should 'enable mfa' do
+            assert @user.reload.mfa_enabled?
           end
         end
-        should 'enable mfa' do
-          assert @user.reload.mfa_enabled?
+
+        context 'when qr-code is expired' do
+          setup do
+            @controller.session[:mfa_seed_expire] = 0
```

Comment:
That cannot make controller think it's expired.

Reply:
Can please explain why would `Time.now.utc > Time.at(Gemcutter::MFA_KEY_EXPIRY + 1.minute).from_now.utc` not work?

Reply:
I meant `(Gemcutter::MFA_KEY_EXPIRY + 1.minute).ago`. sorry.

Reply:
I really want to include `Gemcutter::MFA_KEY_EXPIRY` in tests. Can you please check if http://api.rubyonrails.org/v5.1/classes/ActiveSupport/Testing/TimeHelpers.html#method-i-travel would work.

Reply:
Thanks for your link. I didn't know Rails has provided helpers for 'time traveling' out of the box. I understand what you mean. But since `session[:mfa_seed]` and `session[:mfa_seed_expiry]` is set in `MultifactorAuths#new`, I have to set the expiry time in session manually. So _travelling_ is not needed here. Maybe it's not a good implementation method.